challenger/src/grinding_challenger.rs

```diff
@@ -0,0 +1,36 @@
+use p3_field::PrimeField64;
+use p3_symmetric::CryptographicPermutation;
+
+use crate::{DuplexChallenger, FieldChallenger};
+
+pub trait GrindingChallenger<F: PrimeField64>: FieldChallenger<F> + Clone {
+    // Can be overridden for more efficient methods not involving cloning, depending on the
+    // internals of the challenger.
+    fn grind(&mut self, bits: usize) -> F {
+        for i in 0..F::ORDER_U64 {
+            let witness = F::from_canonical_u64(i);
+            let mut forked = self.clone();
+
+            if forked.check_witness(bits, witness) {
+                self.observe(witness);
```

Comment:
@dlubarov To get the tests to pass, I needed to add these two lines, so that the challenger itself — not just its clone — observes & samples in the proving (to match the verifying). This is obviously not ideal... how do you feel about changing it so that `grind` returns the witness and the new copy of the challenger, which is swapped in?

Reply:
Although I realized, to do that we'd need to pass the new challenger up out of `prove` as well, which would make `prove`'s interface a lot less nice...

Reply:
I think both options seem reasonable, i.e. either

- We can keep these steps in `grind`; if so we can shorten it slightly by calling `check_witness`.
- Or we can say it's the prover's responsibility to call `check_witness` afterward. It's less caller-friendly in a way, but arguably it's nice since the prover's and verifier's challenger code would match more closely.

Reply:
Ohh both good points! Hm the first option feels slightly nicer to me, just having it self-contained